### ISSUE TITLE
[Feature]: Model pruning per-layer sparsity callback for visualization

### ISSUE THREAD

**Issue by Adit-Jain-srm** [Adit-Jain-srm/NightmareNet#510]
## Problem

The compression/pruning module removes model weights but exposes no per-layer sparsity data to callers. Users cannot visualize which layers are most affected or track sparsity progression across cycles.

## Acceptance Criteria

- [ ] `MagnitudePruner` emits per-layer sparsity data via callback after each pruning step
- [ ] Data format: `{layer_name: str, total_params: int, zero_params: int, sparsity_pct: float}`
- [ ] Sparsity data included in pipeline metrics (available in run state)
- [ ] Test: verify callback is called with correct sparsity values for a known model

## Scope

- `nightmarenet/compression/pruner.py` - add sparsity callback
- `tests/test_compression.py` - add sparsity reporting test

> **Important:** Please read CONTRIBUTING.md in full before requesting assignment.

**Comment by Prathvikmehra** [comment]
Hi @Adit-Jain-srm  Please assign this to me 

## Implementation Approach

**Issue Analysis:**
`MagnitudePruner.apply()` in `nightmarenet/compression/pruning.py:33` iterates `model.named_parameters()`, computes a pruning threshold via `kthvalue`, creates a binary mask, and applies `param.mul_(mask)`. Currently it returns only **aggregate** statistics (`{"pruned_params", "total_params", "sparsity"}`) — no per-layer data is emitted.

**Plan:**

### Step 1: Add callback parameter to `apply()`
Extend `MagnitudePruner.apply()` to accept an optional `on_layer_pruned` callback:
```python
def apply(self, model: nn.Module, on_layer_pruned: Optional[Callable] = None) -> dict
```

### Step 2: Emit per-layer data inside the parameter loop
Currently the loop skips bias/embedding params and processes all remaining. Restructure to group params by layer, emit callback after each layer:
```python
layer_data = {}
for name, param in model.named_parameters():
    if param.dim() < 2 or not param.requires_grad:
        continue
    layer_name = ".".join(name.split(".")[:-1])  # strip weight/bias suffix
    # ... compute mask, apply ...
    if layer_name not in layer_data:
        layer_data[layer_name] = {"total_params": 0, "zero_params": 0}
    layer_data[layer_name]["total_params"] += param.numel()
    layer_data[layer_name]["zero_params"] += (mask == 0).sum().item()
    if on_layer_pruned:
        on_layer_pruned(layer_name, total, zero, sparsity_pct)
```

### Step 3: Include sparsity data in pipeline metrics
In `pipeline_runner.py`, pass a callback to `MagnitudePruner.apply()` that stores per-layer sparsity in the run state's `metrics` dict under `per_layer_sparsity`.

### Step 4: Testing
- Create a small mock `nn.Module` with 3-4 layers of known dimensions
- Call `MagnitudePruner.apply()` with a callback that records each invocation
- Verify: callback is called once per layer, sparsity_pct values are correct
- Verify: `total_params == zero_params + non_zero_params` for each layer

**Files changed:**
- `nightmarenet/compression/pruning.py` (add callback parameter and per-layer logic)
- `nightmarenet/pipeline_runner.py` (wire callback in compression/phases)
- `tests/test_compression.py` (add sparsity reporting test)

**Comment by Ayush-0918** [comment]
Hi @Adit-Jain-srm Please assign this to me

**Comment by Adit-Jain-srm** [comment]
Assigned @Ayush-0918
